src/ray/gcs/gcs_server/gcs_health_check_manager.cc

```diff
@@ -0,0 +1,126 @@
+
+#include "ray/gcs/gcs_server/gcs_health_check_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsHealthCheckManager::GcsHealthCheckManager(
+    instrumented_io_context &io_service,
+    std::function<void(const NodeID &)> on_node_death_callback,
+    int64_t initial_delay_ms,
+    int64_t timeout_ms,
+    int64_t period_ms,
+    int64_t failure_threshold)
+    : io_service_(io_service),
+      on_node_death_callback_(on_node_death_callback),
+      initial_delay_ms_(initial_delay_ms),
+      timeout_ms_(timeout_ms),
+      period_ms_(period_ms),
+      failure_threshold_(failure_threshold) {}
+
+GcsHealthCheckManager::~GcsHealthCheckManager() {}
+
+void GcsHealthCheckManager::RemoveNode(const NodeID &node_id) {
+  io_service_.dispatch(
+      [this, node_id]() {
+        auto iter = health_check_contexts_.find(node_id);
+        if (iter == health_check_contexts_.end()) {
+          return;
+        }
+        health_check_contexts_.erase(iter);
+      },
+      "GcsHealthCheckManager::RemoveNode");
+}
+
+void GcsHealthCheckManager::FailNode(const NodeID &node_id) {
+  RAY_LOG(WARNING) << "Node " << node_id << " is dead because the health check failed.";
+  on_node_death_callback_(node_id);
+  health_check_contexts_.erase(node_id);
+}
+
+std::vector<NodeID> GcsHealthCheckManager::GetAllNodes() const {
+  std::vector<NodeID> nodes;
+  for (const auto &[node_id, _] : health_check_contexts_) {
+    nodes.emplace_back(node_id);
+  }
+  return nodes;
+}
+
+void GcsHealthCheckManager::HealthCheckContext::StartHealthCheck() {
+  using ::grpc::health::v1::HealthCheckResponse;
+
+  context_ = std::make_shared<grpc::ClientContext>();
+
+  auto deadline =
+      std::chrono::system_clock::now() + std::chrono::milliseconds(manager_->timeout_ms_);
+  context_->set_deadline(deadline);
+  stub_->async()->Check(
+      context_.get(),
+      &request_,
+      &response_,
+      [this, stopped = this->stopped_, context = this->context_](::grpc::Status status) {
+        if (status.error_code() == ::grpc::StatusCode::CANCELLED) {
+          return;
+        }
+
+        manager_->io_service_.post(
+            [this, stopped, status]() {
+              // Stopped has to be read in the same thread where it's updated.
+              if (*stopped) {
+                return;
+              }
+              RAY_LOG(DEBUG) << "Health check status: " << int(response_.status());
+
+              if (status.ok() && response_.status() == HealthCheckResponse::SERVING) {
+                // Health check passed
+                health_check_remaining_ = manager_->failure_threshold_;
+              } else {
+                --health_check_remaining_;
+                RAY_LOG(WARNING) << "Health check failed for node " << node_id_
+                                 << ", remaining checks " << health_check_remaining_;
+              }
+
+              if (health_check_remaining_ == 0) {
+                manager_->io_service_.post([this]() { manager_->FailNode(node_id_); },
```

Comment:
maybe we don't need to post again? Since we are already in the same io service

Reply:
Remove the node will make 'this' not valid. So run it in another place after this. I can put some comments there.